storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
+			reqLogger.Info("pending volume cannot be marked for deletion: requeue")
+			return reconcile.Result{Requeue: true}, nil
+		}
+		reqLogger.Info("marking volume for deletion")
+		return r.setVolumePhase(volume, storagev1alpha1.VolumeTerminating)
 	}
-	for hostname := range result {
+
+	switch volume.Status.Phase {
+	case "": // Volume without status == freshly created volume.
+		reqLogger.Info("new volume detected")
+		return r.handleNewVolume(request, volume)
+	case storagev1alpha1.VolumePending:
+		reqLogger.Info("volume creation in progress")
+		return r.handlePendingVolume(request, volume)
+	case storagev1alpha1.VolumeAvailable:
+		reqLogger.Info("volume is available")
+		return r.handleAvailableVolume(request, volume)
+	case storagev1alpha1.VolumeFailed:
+		reqLogger.Info("volume is in a failed state")
+		return r.handleFailedVolume(request, volume)
+	case storagev1alpha1.VolumeTerminating:
+		reqLogger.Info("volume destruction in progress")
+		return r.handleTerminatingVolume(request, volume)
+	default:
 		reqLogger.Info(
-			"ping", "hostname", hostname, "result", result[hostname].(bool),
+			"invalid volume state", "Volume.State", volume.Status.Phase,
+		)
+		return r.setVolumeFailed(
+			volume,
+			storagev1alpha1.InternalError,
+			"got invalid state `%s` for volume", volume.Status.Phase,
 		)
 	}
+}
 
-	pv := newPersistentVolumeForCR(instance)
+func (self *ReconcileVolume) handleNewVolume(
+	request reconcile.Request, volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+
+	// Check if a type is specified.
+	if volume.Spec.SparseLoopDevice == nil &&
+		volume.Spec.RawBlockDevice == nil {
+		return self.setVolumeFailed(
+			volume,
+			storagev1alpha1.InternalError,
+			"volume type not found in spec",
+		)
+	}
 
-	// Set Volume instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pv, r.scheme); err != nil {
+	// Check if a PV already exists for this volume.
+	pv, err := self.getPersistentVolume(volume.Name)
+	if err != nil {
+		reqLogger.Error(
+			err, "cannot read PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
 		return reconcile.Result{}, err
 	}
 
-	// Check if this PV already exists
-	found := &corev1.PersistentVolume{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: pv.Name}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new PersistentVolume", "PersistentVolume.Name", pv.Name)
-		err = r.client.Create(context.TODO(), pv)
+	// PV already exists but we don't have a status.
+	if pv != nil {
+		return self.setVolumeFailed(
+			volume, storagev1alpha1.UnavailableError,
+			"PV found but no status: volume has been tampered with",
+			pv.Status.Reason, pv.Status.Message,
+		)
+	}
+
+	return self.setVolumePhase(volume, storagev1alpha1.VolumePending)
+}
+
+func (self *ReconcileVolume) handlePendingVolume(
```

Comment:
This should somehow be broken up. Too many lines between the various `if ... && ...` condition checks, which makes it almost impossible to validate exhaustiveness.

Reply:
Yes, this is just a big WIP commit to have something working (and now it works :slightly_smiling_face: )

There is a lot of duplication/refactor/cleanup to be made, and I'll start today.